pkg/api/v1beta1/types.go

```diff
@@ -397,6 +397,23 @@ type EnvVar struct {
 	Key  string `json:"key,omitempty" description:"name of the environment variable; must be a C_IDENTIFIER; deprecated - use name instead"`
 	// Optional: defaults to "".
 	Value string `json:"value,omitempty" description:"value of the environment variable; defaults to empty string"`
+	// Optional: specify a source for the value of the EnvVar.
+	Source *EnvVarSource `json:"source,omitempty" description:"source for the environment variable's value; cannot be used if value is not empty"`
+}
+
+// EnvVarSource represents a source for the value of an EnvVar.
+// Only one of its members may be specified.
+type EnvVarSource struct {
+	// Selects a field of the pod; only name and namespace are supported.
+	PodField *ObjectFieldSelector `json:"podField,omitempty" description:"selectes a field of the pod; only name and namespace are supported"`
+}
+
+// ObjectFieldSelector selects an APIVersioned field of an object.
+type ObjectFieldSelector struct {
+	// The API version the FieldPath is written in terms of.
+	APIVersion string `json:"apiVersion,omitempty" description="The API version that FieldPath is written in terms of"`
```

Comment:
@bgrant0607 is `ValueFromField` a string?
On Wed, Apr 22, 2015 at 12:30 AM Brian Grant <notifications@github.com>
wrote:

> In pkg/api/v1beta1/types.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#discussion_r28843747>
> :
>
> > @@ -397,6 +397,23 @@ type EnvVar struct {
> >  	Key  string `json:"key,omitempty" description:"name of the environment variable; must be a C_IDENTIFIER; deprecated - use name instead"`
> >  	// Optional: defaults to "".
> >  	Value string `json:"value,omitempty" description:"value of the environment variable; defaults to empty string"`
> > +	// Optional: specify a source for the value of the EnvVar.
> > +	Source *EnvVarSource `json:"source,omitempty" description:"source for the environment variable's value; cannot be used if value is not empty"`
> > +}
> > +
> > +// EnvVarSource represents a source for the value of an EnvVar.
> > +// Only one of its members may be specified.
> > +type EnvVarSource struct {
> > +	// Selects a field of the pod; only name and namespace are supported.
> > +	PodField *ObjectFieldSelector `json:"podField,omitempty" description:"selectes a field of the pod; only name and namespace are supported"`
> > +}
> > +
> > +// ObjectFieldSelector selects an APIVersioned field of an object.
> > +type ObjectFieldSelector struct {
> > +	// The API version the FieldPath is written in terms of.
> > +	APIVersion string `json:"apiVersion,omitempty" description="The API version that FieldPath is written in terms of"`
>
> Same problem as ObjectReference, validation errors, field selectors, and
> other field references.
>
> Sorry I missed the intervening discussion (too much going on), but
> podField and ObjectFieldSelector seem inelegant. ObjectReference was
> sufficient and more extensible. I understand that it seems overly general
> right now, but if we wanted more specific, I'd rather scrap both
> EnvVarSource and ObjectFieldSelector and just put ValueFromField in EnvVar.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739/files#r28843747>
> .
>